File: plugins/plugin-matrix/rust/src/actions/list_rooms.rs
//! List rooms action for Matrix plugin.

use crate::service::MatrixService;
use crate::types::MatrixRoom;
use serde::{Deserialize, Serialize};

/// Result from the list rooms action.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListRoomsResult {
    pub success: bool,
    pub room_count: usize,
    pub rooms: Vec<RoomInfo>,
    pub formatted_text: String,
}

/// Room information for display.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoomInfo {
    pub room_id: String,
    pub name: Option<String>,
    pub alias: Option<String>,
    pub member_count: usize,
    pub is_encrypted: bool,
}

impl From<MatrixRoom> for RoomInfo {
    fn from(room: MatrixRoom) -> Self {
        Self {
            room_id: room.room_id,
            name: room.name,
            alias: room.canonical_alias,
            member_count: room.member_count,
            is_encrypted: room.is_encrypted,
        }
    }
}

/// Execute the list rooms action.
pub async fn execute_list_rooms(service: &MatrixService) -> ListRoomsResult {
    let rooms = service.get_joined_rooms().await;

    let room_infos: Vec<RoomInfo> = rooms.into_iter().map(|r| r.into()).collect();

    // Format room list
    let formatted = if room_infos.is_empty() {
        "Not currently in any rooms.".to_string()
    } else {
        let room_lines: Vec<String> = room_infos
            .iter()
            .map(|room| {
                let name = room
                    .name
                    .as_ref()
                    .or(room.alias.as_ref())
                    .map(|s| s.as_str())
                    .unwrap_or(&room.room_id);
                let encrypted = if room.is_encrypted { " 🔒" } else { "" };
                format!("• {} ({} members){}", name, room.member_count, encrypted)
            })
            .collect();

        format!(
            "Joined {} room(s):\n\n{}",
            room_infos.len(),
            room_lines.join("\n")
        )
    };

    ListRoomsResult {
        success: true,
        room_count: room_infos.len(),
        rooms: room_infos,
        formatted_text: formatted,
    }
}

File: plugins/plugin-matrix/typescript/src/actions/listRooms.ts
/**
 * List rooms action for Matrix plugin.
 */

import type { Action, ActionResult, HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State } from "@elizaos/core";
import { MatrixService } from "../service.js";
import { MATRIX_SERVICE_NAME } from "../types.js";

export const listRooms: Action = {
  name: "MATRIX_LIST_ROOMS",
  similes: [
    "LIST_MATRIX_ROOMS",
    "SHOW_ROOMS",
    "GET_ROOMS",
    "MY_ROOMS",
  ],
  description: "List all Matrix rooms the bot has joined",

  validate: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State
  ): Promise<boolean> => {
    return message.content.source === "matrix";
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult | undefined> => {
    const matrixService = (await runtime.getService(MATRIX_SERVICE_NAME)) as MatrixService | undefined;

    if (!matrixService || !matrixService.isConnected()) {
      if (callback) {
        await callback({ text: "Matrix service is not available.", source: "matrix" });
      }
      return { success: false, error: "Matrix service not available" };
    }

    const rooms = await matrixService.getJoinedRooms();

    // Format room list
    const roomList = rooms.map((room) => {
      const name = room.name || room.canonicalAlias || room.roomId;
      const members = `${room.memberCount} members`;
      const encrypted = room.isEncrypted ? " (encrypted)" : "";
      return `- ${name} (${members})${encrypted}`;
    });

    const responseText =
      rooms.length > 0
        ? `Joined ${rooms.length} room(s):\n\n${roomList.join("\n")}`
        : "Not currently in any rooms.";

    if (callback) {
      await callback({
        text: responseText,
        source: message.content.source as string,
      });
    }

    return {
      success: true,
      data: {
        roomCount: rooms.length,
        rooms: rooms.map((r) => ({
          roomId: r.roomId,
          name: r.name,
          alias: r.canonicalAlias,
          memberCount: r.memberCount,
          isEncrypted: r.isEncrypted,
        })),
      },
    };
  },

  examples: [
    [
      {
        name: "{{user1}}",
        content: { text: "What rooms are you in?" },
      },
      {
        name: "{{agent}}",
        content: {
          text: "I'll list the rooms I've joined.",
          actions: ["MATRIX_LIST_ROOMS"],
        },
      },
    ],
  ],
};
